### PR TITLE
Billions - miliardy

### PR DESCRIPTION
**Pull request by ipepe**
I added billions support

### CODE REVIEW

**Comment by ipepe** [comment]
I don't know why travis is reporting errors? Could anyone tell me what should I fix?

**Comment by wpiekutowski** [comment]
Thanks for the PR. I'll take a look at the travis issue and code itself, but first please fix the following cosmetic issues:
- use only English
- close the previous PR
- squash all code climate commits together
- squash changelog commits together
- remove the merge commit

**Comment by ipepe** [comment]
Well in my commits I added only billions support. That is why I added it as separate PR

**Comment by ipepe** [comment]
@wpiekutowski 
I did all You asked about, except closing PR because I'm not owner of it.

**Comment by wpiekutowski** [comment]
Thank you very much. I probably won't be able to take care of this until
Monday though, so please be patient.

**Comment by ipepe** [comment]
@wpiekutowski How it is going? Becasue we would like to update Gem without pointing to our private repositories

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 *.gem
 Gemfile.lock
 .bundle
+.idea
```
Comment (wpiekutowski) [inline .gitignore]: Missing new line.

File: .ruby-version [hunk 2]
```diff
@@ -1 +1 @@
-2.1.5
+2.1.5
```
Comment (wpiekutowski) [inline .ruby-version]: Missing new line.

File: README.md [hunk 3]
```diff
@@ -26,6 +26,35 @@ Or install it yourself as:
     PolishNumber.translate(34, :currency => :PLN) #=> trzydzieści cztery złote
     PolishNumber.translate(12, :currency => :PLN) #=> dwanaście złotych
 
+## Cents format
+
+    PolishNumber.translate(34.11) #=> trzydzieści cztery i jedenaście setnych
```
Comment (wpiekutowski) [inline README.md:31]: Do you actually use floats to store currency amounts?

**Comment by wpiekutowski** [comment]
@ipepe you've squashed all commits together. This means that @simcha authorship information is lost. @simcha is that ok with you?

File: README.md [hunk 4]
```diff
@@ -26,6 +26,35 @@ Or install it yourself as:
...
+    PolishNumber.translate(0.10, :currency => :PLN) #=> dziesięć groszy
+    PolishNumber.translate(5.10, :currency => :PLN, :cents => :no ) #=> pięć złotych
+    PolishNumber.translate(5, :currency => :PLN, :cents => :words ) #=> pięć złotych zero groszy
+    PolishNumber.translate(5, :currency => :PLN, :cents => :digits ) #=> pięć złotych 00/100
+
+## Adding new currencies
+
+    PolishNumber.add_currency(:COWS, { :one => 'krowa', :few => 'krowy',
+      :many => 'krów', :gender => :she,
+      :one_100 => 'ser', :few_100 => 'sery',
```
Comment (wpiekutowski) [inline README.md:42]: `:one_cent`, `:few_cents`, `:many_cents` seem to be easier to understand.

**Comment by wpiekutowski** [comment]
I'd like to work with both of you to merge this PR. I think it could be useful for many people working on projects with Polish language.

File: README.md [hunk 5]
```diff
@@ -26,6 +26,35 @@ Or install it yourself as:
...
     PolishNumber.translate(12, :currency => :PLN) #=> dwanaście złotych
 
+## Cents format
+
+    PolishNumber.translate(34.11) #=> trzydzieści cztery i jedenaście setnych
+    PolishNumber.translate(30.01, :currency => :PLN) #=> trzydzieści złote i jeden grosz
+    PolishNumber.translate(0.10, :currency => :PLN) #=> dziesięć groszy
+    PolishNumber.translate(5.10, :currency => :PLN, :cents => :no ) #=> pięć złotych
+    PolishNumber.translate(5, :currency => :PLN, :cents => :words ) #=> pięć złotych zero groszy
+    PolishNumber.translate(5, :currency => :PLN, :cents => :digits ) #=> pięć złotych 00/100
```
Comment (wpiekutowski) [inline README.md]: `:cents` option is a good idea. Can we rename this value to something that more closely represents true meaning? For example `:fraction`?

File: README.md [hunk 6]
```diff
@@ -26,6 +26,35 @@ Or install it yourself as:
     PolishNumber.translate(34, :currency => :PLN) #=> trzydzieści cztery złote
     PolishNumber.translate(12, :currency => :PLN) #=> dwanaście złotych
 
+## Cents format
+
+    PolishNumber.translate(34.11) #=> trzydzieści cztery i jedenaście setnych
+    PolishNumber.translate(30.01, :currency => :PLN) #=> trzydzieści złote i jeden grosz
+    PolishNumber.translate(0.10, :currency => :PLN) #=> dziesięć groszy
+    PolishNumber.translate(5.10, :currency => :PLN, :cents => :no ) #=> pięć złotych
```
Comment (wpiekutowski) [inline README.md]: Can we rename this value to `:none` please?

File: README.md [hunk 7]
```diff
@@ -26,6 +26,35 @@ Or install it yourself as:
...
+    PolishNumber.translate(5, :currency => :PLN, :cents => :words ) #=> pięć złotych zero groszy
+    PolishNumber.translate(5, :currency => :PLN, :cents => :digits ) #=> pięć złotych 00/100
+
+## Adding new currencies
+
+    PolishNumber.add_currency(:COWS, { :one => 'krowa', :few => 'krowy',
+      :many => 'krów', :gender => :she,
```
Comment (wpiekutowski) [inline README.md]: Can we use masculine, feminine and neuter for gender value?

File: dev_console.sh [hunk 8]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+# Run the console with all the code loaded
+pry -I "./lib" -r "./lib/polish_number"
```
Comment (wpiekutowski) [inline dev_console.sh]: Missing new line.

File: README.md (hunk 6 above)
Comment (wpiekutowski) [inline README.md]: Unneeded space before `)`

File: README.md [hunk 9]
```diff
@@ -26,6 +26,35 @@ Or install it yourself as:
     PolishNumber.translate(34, :currency => :PLN) #=> trzydzieści cztery złote
     PolishNumber.translate(12, :currency => :PLN) #=> dwanaście złotych
 
+## Cents format
+
+    PolishNumber.translate(34.11) #=> trzydzieści cztery i jedenaście setnych
+    PolishNumber.translate(30.01, :currency => :PLN) #=> trzydzieści złote i jeden grosz
+    PolishNumber.translate(0.10, :currency => :PLN) #=> dziesięć groszy
+    PolishNumber.translate(5.10, :currency => :PLN, :cents => :no ) #=> pięć złotych
+    PolishNumber.translate(5, :currency => :PLN, :cents => :words ) #=> pięć złotych zero groszy
```
Comment (wpiekutowski) [inline README.md]: Unneeded space before `)`

File: README.md (hunk 5 above)
Comment (wpiekutowski) [inline README.md]: Unneeded space before `)`

File: README.md [hunk 10]
```diff
@@ -26,6 +26,35 @@ Or install it yourself as:
...
+    PolishNumber.translate(5, :currency => :PLN, :cents => :digits ) #=> pięć złotych 00/100
+
+## Adding new currencies
+
+    PolishNumber.add_currency(:COWS, { :one => 'krowa', :few => 'krowy',
```
Comment (wpiekutowski) [inline README.md:40]: Unneeded space after `{`. Please improve formatting of this part.

Comment (wpiekutowski) [inline README.md:40]: This isn't a currency. It's a unit. Can we keep `:currency` in `.translate` for backwards compatibility? Can we add there `:unit` as an alias to `:currency`? Can we rename this method to `.add_unit`?

File: lib/polish_number.rb [hunk 11]
```diff
@@ -21,86 +21,181 @@ module PolishNumber
 
   MILLIONS = {:one => 'milion', :few => 'miliony', :many => 'milionów'}
 
+  BILLIONS = {:one => 'miliard', :few => 'miliardy', :many => 'miliardów'}
+
+  CENTS = [:auto, :no, :words, :digits]
+
   CURRENCIES = {
```
Comment (wpiekutowski) [inline lib/polish_number.rb:28]: Again `UNITS` would make much more sense, if you want to have support for arbitrary words, like in the next line.

File: lib/polish_number.rb [hunk 12]
```diff
@@ -21,86 +21,181 @@ module PolishNumber
...
-    :PLN => {:one => 'złoty', :few => 'złote', :many => 'złotych'}
+    :NO => {:one => '', :few => '', :many => '',
```
Comment (wpiekutowski) [inline lib/polish_number.rb]: Is NO a shortcut from NUMBER?

File: lib/polish_number.rb [hunk 13]
```diff
@@ -21,86 +21,181 @@ module PolishNumber
...
-      raise ArgumentError, 'number should be in 0..999999999 range'
+    digits = formatted_number.chars.map { |char| char.to_i }
+    result = process_1_999999999999(digits[0..11], options, number, currency)
+
+    process_99_0(result, digits, options, formatted_number[-2..-1], currency)
+
+  end
+
+  def self.add_currency(name, hash)
+    CURRENCIES[name]=hash
```
Comment (wpiekutowski) [inline lib/polish_number.rb:79]: This isn't thread-safe. Perhaps we should use https://github.com/ruby-concurrency/thread_safe or immutable data structure?

**Comment by wpiekutowski** [comment]
I see some other minor formatting issues, but we can tackle that after we figure out some more important things:
- using floats for currency amounts
- thread safety
- naming things

@ipepe I'd like to push here a fix to travis CI issue. Could you give me access to your fork?

**Comment by ipepe** [comment]
I fixed some formatting and naming issues that You addressed. But I can't help You more in this matter. Also I gave You permission to my repository

**Comment by wpiekutowski** [comment]
Thanks!

**Comment by wpiekutowski** [comment]
I'm closing this because multiple issues with this PR were never fixed.

**Pull request closed without merging**